### ISSUE TITLE
[FEATURE] Ensure that we can support DBR 14.3LTS

### ISSUE THREAD

**Issue by dannymeijer** [Nike-Inc/koheesio#33]
## Is your feature request related to a problem? Please describe.
N/A

## Describe the solution you'd like

We should add support for [DBR 14.3 LTS](https://docs.databricks.com/en/release-notes/runtime/14.3lts.html)

This means we need compatibility with:
- [x] Apache Spark: 3.5.0
- [x] Python: 3.10.12  (already covered)
- [x] Delta Lake: 3.1.0

Additionally, we need to look at how Spark Connect changes things for us. 
Any reference we have to JVM directly, we should investigate. 
Only Shared cluster mode is affected according to docs.

## Describe alternatives you've considered
N/A

## Additional context
N/A

**Comment by mikita-sakalouski** [comment]
The whole idea was to introduce the internal koheesio spark session to be able to provide easy switch between remote and local modes. 

Also if I'm not wrong pydantic is checking SparkSession type based on the full import path and with remote spark session it is imported from a different path, at least it was in a such way sometime ago

**Comment by pariksheet** [comment]
https://docs.databricks.com/en/dev-tools/databricks-connect/python/limitations.html

**Comment by pariksheet** [comment]
Not available on Databricks Connect for Databricks Runtime 13.3 LTS and below:

Streaming foreachBatch

Long queries over 3600 seconds -- not sure what does this mean.

Not available:

Databricks Utilities: credentials, library, notebook workflow, widgets  -- not sure what does this mean.

SparkContext -- No jvm operations

Changing the log4j log level through SparkContext -- need to check the code

**Comment by pariksheet** [comment]
Run the unit test locally with spark-connect remote instant.

**Comment by pariksheet** [comment]
Check how to manage SparkSession and DatabricksSession.

**Comment by femilian-6582** [comment]
### Long queries over 3600 seconds 

It has probably to do with the parameter **STATEMENT_TIMEOUT** that sets a SQL statement timeout in seconds. This parameter is set at the Session level. So, concerning the Spark Connect, where your entry point is the DatabricksSession:

```
from databricks.connect import DatabricksSession

spark = DatabricksSession.builder.getOrCreate() 
```

it seems this parameter is set to 3600 and cannot be modified. I still couldn't find where this happens though. 

https://docs.databricks.com/en/sql/language-manual/parameters/statement_timeout.html 

https://docs.databricks.com/en/sql/language-manual/sql-ref-parameters.html

**Comment by riccamini** [comment]
I have added details in here related to foreachbatch function: https://github.com/Nike-Inc/koheesio/issues/56

If you prefer collecting everything here I will copy paste the comment and close the issue.

One additional point that I do not see in the list is `Dataframe.rdd` which is being used in some tests

**Comment by pariksheet** [comment]
- [ ] Delta merge through Delta API not possible. (only supported from Spark 4.0 and [Delta 4.0](https://github.com/delta-io/delta/releases/tag/v4.0.0rc1]))
 - [ ] Delta Merge through SPARK SQL still works
 - [ ] How to support 3.x and 4.x spark/delta versions

**Comment by pariksheet** [comment]
There is a way to check Spark Session is remote or native.

we should introduce api/function to get spark session flag and check against the specific APIs e.g. delta merge /snowflake and raise the exception.

**Comment by pariksheet** [comment]
-- use snowflake-connector-python instead of spark._jvm

**Comment by dannymeijer** [comment]
All of these should be addressed as part of release 0.9.0 (currently in pre-release). Please verify your usecases accordingly so we can proceed with the release.
